### ISSUE TITLE
Elementum not work on Android when change Kodi default data folder

### ISSUE THREAD

**Issue by alfonsu** [elgatito/plugin.video.elementum#920]
Hello :)
I need help to solve the following problem:
When I change the default location from where kodi loads user data on android 11+ Android/data/org.xbmc.kodi/files/.kodi to data/org.xbmc.kodi/files/.kodi because from android 11+ it is no longer accessible next to the Android/data folder all the plugins I use work only without Eelementum, it gives the following error Daemon error:b"FileNotFoundError(2,'No such file or directory)". 
Obviously the problem comes from changing the default location, but the question is how can I fix it. 
To change the location I follow Kodi's instructions and have created an xbmc_env.properties file in /storage/emulated/0 with the following content xbmc.data=/storage/emulated/0/data/org.xbmc.kodi/files and this works for me.

**Comment by antonsoroko** [comment]
@alfonsu Hi.
first, not sure why you need to change userdata location, people with android 11+ run elementum w/o this, see https://github.com/elgatito/plugin.video.elementum/issues/669

if by some reason you still want to change directory, make sure that you also **changed all paths in elementum settings**.

by default elementum uses "special" (relative) folders, see https://kodi.wiki/view/Special_protocol and 
https://github.com/elgatito/elementum/blob/2613fa116380687407247d6464fb87f617fe6288/api/cmd.go#L85
https://github.com/elgatito/plugin.video.elementum/blob/7680e0bd365644a9cdae2e6b50bb55868f27593c/resources/settings.xml#L53

also, make log file like issue template (that you ignored :-) ) said and read it - maybe it will have the answer, if not - upload it and paste link here.

**Comment by alfonsu** [comment]
@antonsoroko 
I need to change the default folder /Android/data because from android 11+ it is no longer accessible in an easy way and I have a lot of add-ons where I enter usernames and passwords and I want to just transfer them without a lot of headaches. All my plugins work after changing the path, only Elementum doesn't. Interestingly, up to and including Android 10, everything works without a problem. Elementum loads from an external folder outside of Android/data without errors. I guess it all comes from the limitations in Android 11+. Now I will share the two log files
one is from the /Android/data folder where Elementum starts without a problem and the second is from my external folder from where there is a problem.
[kodi1.log](https://github.com/elgatito/plugin.video.elementum/files/11713076/kodi1.log)
[kodi2.log](https://github.com/elgatito/plugin.video.elementum/files/11713078/kodi2.log)

**Comment by antonsoroko** [comment]
looks like you created log files without debugging enabled (guide says how to enable it both in kodi and elementum).
so i can't tell you much.

the only thing that i can see is that elementum fails to open its log file:
in 1st log it is `/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/temp/elementum/elementum.log`
in 2nd log it is `/storage/emulated/0/data/org.xbmc.kodi/files/.kodi/temp/elementum/elementum.log`

maybe check if this new path exists (and if it is writable from within the kodi, if you can use some internal kodi's file manager).

https://github.com/elgatito/plugin.video.elementum/blob/7680e0bd365644a9cdae2e6b50bb55868f27593c/resources/site-packages/elementum/daemon.py#L401

**Comment by alfonsu** [comment]
@antonsoroko 
It's my fault that I forgot to put the debug log. Here are the new files.
[kodi1.log](https://github.com/elgatito/plugin.video.elementum/files/11730774/kodi1.log)
[kodi2.log](https://github.com/elgatito/plugin.video.elementum/files/11730775/kodi2.log)

**Comment by antonsoroko** [comment]
@alfonsu i think you still forgot to enable debug in elementum itself.
anyway, what about my question about path?

**Comment by alfonsu** [comment]
@antonsoroko 
Yes I really forgot to enable debug in elementum. Now I upload the files again with debug enabled. The daemon.py file is the same in both places and contains line 401
for line in follow(io.open(log_path, 'r', encoding='utf8'), delay=0.5, selector=proc.is_alive):
[kodi1.log](https://github.com/elgatito/plugin.video.elementum/files/11735532/kodi1.log)
[kodi2.log](https://github.com/elgatito/plugin.video.elementum/files/11735534/kodi2.log)

**Comment by antonsoroko** [comment]
@antonsoroko "about path" i meant answer my question from https://github.com/elgatito/plugin.video.elementum/issues/920#issuecomment-1586214539

**Comment by antonsoroko** [comment]
@alfonsu  
1. "about path" i meant do the check from my comment https://github.com/elgatito/plugin.video.elementum/issues/920#issuecomment-1586214539

>check if this new path exists (and if it is writable from within the kodi, if you can use some internal kodi's file manager).

2. i still do not see debug messages from elementum in log, did you restart kodi?
there should be lines like
```
... warning <general>: [plugin.video.elementum] INFO  main         ▶ main             ^[[0mStarting Elementum daemon
... warning <general>: [plugin.video.elementum] INFO  main         ▶ main             ^[[0mVersion: 0.1.87 LibTorrent: 1.1.14.0 Go: go1.19.4, Threads: 8
```

**Comment by alfonsu** [comment]
@antonsoroko 
The path exists and can be writable by KODI. Debug mode is started and then Kodi is restarted, having previously deleted the old log files.
![Screenshot_2023-06-13-17-11-44-677](https://github.com/elgatito/plugin.video.elementum/assets/90715350/bbc39cf1-ce4b-4620-8381-a5c8153bc08a)
![Screenshot_2023-06-13-17-12-26-660](https://github.com/elgatito/plugin.video.elementum/assets/90715350/1425aca9-4639-40b0-838b-91e21670cc93)

**Comment by antonsoroko** [comment]
@alfonsu 

>The path exists and can be writable by KODI. 

elementum says otherwise (or i misread the code). 
Was `/storage/emulated/0/data/org.xbmc.kodi/files/.kodi/temp/elementum/elementum.log` file have been created?

**Comment by alfonsu** [comment]
@antonsoroko 
no such elementum.log file in both locations
/storage/emulated/0/data/org.xbmc.kodi/files/.kodi/temp/elementum/
/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/temp/elementum/

**Comment by antonsoroko** [comment]
@alfonsu 
i missed the important line in the log:
`2023-06-13 16:30:03.642 T:18356 warning <general>: [plugin.video.elementum] Shared library execution stopped with code 1`

for new android author made a workaround, instead of binary he uses library (see end of #669), but looks like if library failed to start then there is no log messages that we can use to see the issue.

https://github.com/elgatito/plugin.video.elementum/commit/18916c1065f67619efc8c032ecfd7d0c75ffed6e#diff-71ebf61bf8dc7b8a527700b3aaf2c64e628b2074cd503d890e1c005a311776b5R350

with this new approarch, looks like you need to use `adb logcat` to see the Android logs, there are guides in internet.

**Comment by elgatito** [comment]
@antonsoroko I've seen the log, but I can't see the reason for getting a segfault.

On start of an app we try to write `.writable` file (and remove it afterwards). If provided path is read only - we will fail. 
But I think it will fail to start with non-memory storage selected. 

@alfonsu Can you try to select memory storage and restart Kodi?

**Comment by alfonsu** [comment]
@elgatito 
I want to use the built-in memory for storage, I am not comfortable with external memory.
I tried changing the path in the config files as suggested by @antonsoroko
TorrentsPath: "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/temp/elementum_torrents",
LibraryPath: "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/temp/elementum_library",
but it didn't help.
However, if I delete xbmc_env.properties, open Kodi it loads as brand new, transfer with built-in file manager everything from folder /storage/emulated/0/data/org.xbmc.kodi/files/.kodi to special://home, reboot kodi and then it works with all my settings from the default android/data/org.xbmc.kodi/ folder and elementum works.
The problem apparently comes from Android 11+ itself, because I tried all versions up to and including Android 10, and there Kodi loads easily from /storage/emulated/0/data/org.xbmc.kodi/ and elementum has no problems.
My idea is where and how can I transfer the data folder to kodi so that I can easily access it, because from android 11+ the Android/data folder is very difficult to access.

**Comment by antonsoroko** [comment]
@alfonsu last attempt: while using `xbmc_env.properties` try to write to `/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/temp/elementum_torrents` and to `/storage/emulated/0/data/org.xbmc.kodi/files/.kodi/temp/elementum_torrents` from Kodi. 
Yes, you said that "The path exists and can be writable by KODI" but you have not said **how** you checked that, so I am not sure. Maybe this little experiment will give us some clues. If write will fail - please also gather `adb logcat`.

**Comment by elgatito** [comment]
@alfonsu Aaaahhh.. So you use this - https://kodi.wiki/view/HOW-TO:Change_data_location_for_Android 
To change your data location for Kodi.
And Elementum is not picking that change, right?

@StefanIlchev that is the thing you were mentioning?

**Comment by StefanIlchev** [comment]
@elgatito, @alfonsu, the problem is that you can't lock files in shared storage. For more information see https://github.com/StefanIlchev/ElementumService/issues/20.

Also, in the Android 14 beta they've extended this to all external storage including the apps' and if this doesn't change Elementum will not work on Android 14 and up.

I've solved this by moving the files to internal storage and was planning to release a new version once Android 14 came out, but made another version still targeting Android 13 since as a side effect it makes it possible to change data location. See [ElementumService](https://github.com/StefanIlchev/ElementumService).

**Comment by alfonsu** [comment]
@StefanIlchev  @elgatito @antonsoroko 
Yes Yes Yes, Working :)

**Comment by elgatito** [comment]
@StefanIlchev Was playing little bit with this on Android 13 emulator. 

If I set `xbmc_env.properties` to point into `/sdcard/Download/.kodi` (for example), then it allows writing files in that location, but not locking them and not running something from there.

If I use `/data/user/0/org.xbmc.kodi/cache/apk/assets` (which is a `special://xbmcbin/` location), then I can run shared library from there.

And with `/sdcard/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.elementum/` I can write and lock files.

So what is the change with Android 14? What is the correct folder Elementum should use to store binary/shared? When to store application files, that require locking?

**Comment by StefanIlchev** [comment]
The change in Android 14 is that you can no longer lock files in `/sdcard/Android/data/org.xbmc.kodi/` as well and not just "outside". You can still lock files in `/data/user/0/org.xbmc.kodi/`.

**Comment by elgatito** [comment]
@StefanIlchev Do you see anything bad if we move all the Elementum's data into `/data/user/0/org.xbmc.kodi/`? Databases, etc?

**Comment by StefanIlchev** [comment]
@elgatito preferably not directly in `/data/user/0/org.xbmc.kodi/` but in `/data/user/0/org.xbmc.kodi/files/Elementum/` for example. Also, the path might be different for different devices or if the app is installed on removable storage, so you should get it from `special://xbmc`. And no, I don't see an issue with moving the databases there. That's where they would be if this was an Android app and not a Kodi addon.

**Comment by alfonsu** [comment]
@elgatito @StefanIlchev 
The only drawback of Elementum Service is that when you install for example version 0.1.87.apk and plugin.video.elementum-android_client.zip is automatically updated from the Elementum repository it stops working until you install the same version ElementumService.apk as happened to me today with the update to 0.1.89.

**Comment by StefanIlchev** [comment]
@alfonsu, auto-updating of the app was broken, but it was fixed in 0.1.890002. Following updates should be fine.

**Comment by alfonsu** [comment]
@StefanIlchev 
Super :)

**Comment by elgatito** [comment]
I've landed v0.1.90, it now allows using custom profile path. 
Was testing on Android 14 emulator.

@alfonsu Can you check it?

**Comment by alfonsu** [comment]
@elgatito 
v0.1.90 Not Working
v0.1.91 Working Without Elementum Service but can't access elementum repository in kodi

**Comment by elgatito** [comment]
@alfonsu What do you mean by not able to access Elementum repository? Can you show a log?

**Comment by alfonsu** [comment]
@elgatito 
![Screenshot_2023-07-05-13-52-26-593](https://github.com/elgatito/plugin.video.elementum/assets/90715350/30448265-5b6c-42cd-a56d-26c3fd37788d)
![Screenshot_2023-07-05-13-52-34-576](https://github.com/elgatito/plugin.video.elementum/assets/90715350/039760de-592b-468c-9037-0e405aa22f4a)
With Clean Kodi data and only install plugin.video.elementum-0.1.91.android_arm.zip

**Comment by elgatito** [comment]
What is in the log? To act as a repository Elementum should be running at that time.

**Comment by alfonsu** [comment]
Elementum Working and Start with Kodi without problems
[adb-logcat.txt](https://github.com/elgatito/plugin.video.elementum/files/11957373/adb-logcat.txt)

**Comment by elgatito** [comment]
@alfonsu Repository is also working?

**Comment by alfonsu** [comment]
1.I install Kodi
2.I install plugin.video.elementum-0.1.91.zip
3.Once plugin.video.elementum is installed it automatically starts and adds the elementum repository
4.Then I want to go into the elementum repository and install it from there so it can update automatically for the future but when I try to do that it gives me an error (could not connect to repository)
I tested it on Windows 10/ Android 10/ Android 11
it's the same for everyone

**Comment by elgatito** [comment]
Can you share kodi log file after you tried to install something from repository?

**Comment by alfonsu** [comment]
Here is the Windows log
[kodi.log](https://github.com/elgatito/plugin.video.elementum/files/11958637/kodi.log)

**Comment by ramiro** [comment]
@alfonsu per your log you most probably were victim of #932
```
Repository add-on repository.elementum uses plain HTTP for add-on downloads in path http://:65220:65220/repository/elgatito/ - this is insecure and will make your Kodi installation vulnerable to attacks if enabled!
```
which @elgatito fixed in 0.1.92.

You might want to check custom profile path implementation available since v0.1.90 and report back.
